Update IDNA encoding to 2008 spec

StephenBrown2 opened the pull request:
Resolves #150 and adds tests based on deviations from 2003 spec: https://unicode.org/reports/tr46/#Deviations

This also enables Unicode® Technical Standard #46 to normalize capital letters and such, and a test based on the example given in the [module documentation](https://github.com/kjd/idna#compatibility-mapping-uts-46).

Since the idna encode() function chokes on colons, I've split and rejoined the authority string and added to the tests to make sure the port stays as expected.

yeraydiazdiaz commented on tests/models/test_url.py:
```diff
@@ -1,12 +1,41 @@
+    assert url.port == 444
+
+    url = URL("https://نامه‌ای.com:4433")
+    assert url == URL("https://xn--mgba3gch31f060k.com:4433") # IDNA 2008
```
Nitpick: I would drop the `IDNA 2008` comments 🙂

StephenBrown2 replied:
Certainly. I think I will also parameterize the test to get better ids.

yeraydiazdiaz commented on httpx/models.py:
```diff
@@ -93,7 +94,11 @@ def __init__(
+            # idna.encode raises InvalidCodepoint when encountering a colon, so split
+            # host and port in case the latter is specified and rejoin after encoding
+            host_port = self.components.authority.split(":")
+            host_port[0] = idna.encode(host_port[0], uts46=True).decode("ascii")
```
Reference on the UTS 46 flag https://github.com/kjd/idna#compatibility-mapping-uts-46, maybe we should include it in a comment?

sethmlarson replied:
Let's drop the uts46 flag in favor of strict=True and std3_rules=True per RFC 3986

StephenBrown2 replied:
Do we want to fail on things like capitalization? I feel we don't necessarily want to be strict, but too much assistance can also be bad.

e.g. This will fail:
```
idna.core.InvalidCodepoint: Codepoint U+004B at position 1 of 'Königsgäßchen' not allowed
```

StephenBrown2 replied:
Ended up using rfc3986.IRIReference.encode() instead and bypassing this entirely.

sethmlarson commented on httpx/models.py:
```diff
@@ -93,7 +94,11 @@ def __init__(
-            idna_authority = self.components.authority.encode("idna").decode("ascii")
+            # idna.encode raises InvalidCodepoint when encountering a colon, so split
+            # host and port in case the latter is specified and rejoin after encoding
+            host_port = self.components.authority.split(":")
```
I think there's a way we can rely on rfc3986 to hand us the host portion of a URI. The current implementation would choke on IPv6 addresses.

StephenBrown2 replied:
This is true, I had also assumed that authority would always be a fqdn/hostname. Getting just the host portion to start with would be best.

StephenBrown2 commented:
Looks like rfc3986.IRIReference will in fact do exactly what we want, including handling colons in the authority. I've rebased on master to resolve the conflict with hsts... and switched to using `rfc3986.api.iri_reference(url).encode()` which will handle the IDNA Domain and return a URLReference object as expected.

Ref: convenience function - https://github.com/python-hyper/rfc3986/blob/df21c4b09078580c1cb10c728dc18c3111fcf603/src/rfc3986/api.py#L41
Relevant bit of IRIReference.encode() - https://github.com/python-hyper/rfc3986/blob/df21c4b09078580c1cb10c728dc18c3111fcf603/src/rfc3986/iri.py#L97-L129

tomchristie commented on httpx/models.py:
```diff
@@ -87,17 +87,14 @@ def __init__(
     ) -> None:
         if isinstance(url, rfc3986.uri.URIReference):
             self.components = url
+        elif isinstance(url, rfc3986.iri.IRIReference):
```
I don't think we should be exposing extra types as part of this pull request.

(In fact I don't think we should be exposing `rfc3986.uri.URIReference` as part of our public interface in the first place - it'd be much better if we could keep that as an internal implementation detail, in order to allow us to completely switch implementations if needed.)

tomchristie replied:
See #170

tomchristie commented:
I guess we *might* also choose to switch our dependency listing to `rfc3986[idna]`, instead of listing the two packages seperately?

That'd stop devs from having to do a "wait where are you actually using the `idna` package?" double take.

tomchristie commented:
Yas - good stuff!
